site/content/blog/2020-02-10-maintaining-consistency-in-codebases-with-go-vet.md

```diff
@@ -0,0 +1,189 @@
+---
+title: Maintaining consistency in codebases with Go Vet
+slug: maintaining-consistency-in-codebases-with-go-vet
+date: 2020-02-10
+categories:
+    - "go"
+author: Jesús Espino
+github: jespino
+community: jesus.espino
+---
+
+Maintaining success in a large open-source project is one of the key objectives
+of Mattermost. We have hundreds of contributors and we want to create a project
+that could serve as a model in the Go community. Having said that, following
+idiomatic Go principles is the thing that we care most while maintaining our
+code consistency. For this specific task, we utilized `go vet` and with this
+blog post, I would like to explain how we pushed the limits of this tool by
+extending it.
+
+The main restriction of `go vet` is that there are a limited number of absolute
+truths about what is right and what is wrong and so the `go vet` checks are very
+general.
+
+Although `go vet` has great use in common checks, having domain-specific or even
+company-specific checks is inevitable to maintain a project at our scale. In
+Mattermost we have a way of doing certain things like logging, test assertions
+and adding license headers to the source files. We try our best to keep code
+consistent therefore we work hard to avoid reintroducing old patterns in the
+code.
+
+I think the best way to explain it is with an example.
+
+Some time ago, we re-designed our logging implementation and I think it would
+be a great example to showcase our work about maintaining consistency. While
+migrating to the new approach it didn't happen with snapping a finger hence, we
+observed that the old way of logging was making its way to the new PRs. Also in
+such cases, the old pattern was one of the obvious approaches that you can
+follow to present data in the logs.
+
+Let's dive into the topic, I will try my best to explain how we extended the go
+vet tool by adding our own specific checks.
+
+A good starting point would be the [check](https://github.com/mattermost/mattermost-govet/blob/master/structuredLogging/structuredLogging.go)
+we added so that we could avoid any string building with `fmt.Sprintf` calls as
+part of the calls to our log library. With that check implemented we were able
+to detect all the cases in the code where we were doing not-structured logging
```

Comment:
```suggestion
to detect all the cases in the code where we were doing not-structured logging
```
Would "unstructured" work as well?

Reply:
I think this is more clear because put the emphasis on the idea of is not "structured" logging.